tests/test_math_kvecgenerator.cc

```diff
@@ -0,0 +1,339 @@
+      // before and after scaling. We are only using powers of two to
+      // eliminate unwanted behavior due to rounding errors.
+      double scales[2] = {0.5, 2};
+      for (auto & scale : scales) {
+        math::Kvectors kvectors_scaled(cutoff * scale, cell / scale);
+        int nvec_scaled = kvectors_scaled.get_numvectors();
+        Eigen::MatrixXd kvecs_scaled = kvectors_scaled.get_kvectors() / scale;
+        BOOST_CHECK_EQUAL(nvec, nvec_scaled);
+
+        // Check that the obtained vectors agree using squared norm
+        if (nvec == nvec_scaled) {
+          for (int ik = 0; ik < nvec; ik++) {
+            normsq = (kvecs.row(ik) - kvecs_scaled.row(ik)).squaredNorm();
+            BOOST_CHECK_LT(normsq, cutoff * cutoff * 1e-14);
+          }  // for all k-vectors (of this cell)
+        }    // if number of vectors agree
+      }      // for all scale factors
+    }        // for each cell
+  }          // end test case
+
+  /* ---------------------------------------------------------------------- */
+  /**
+   * @brief Test correct behavior under rotations
+   *
+   * If we rotate a (real space) cell and compute its k-vectors, the
+   * obtained k-vectors should be the same as the rotation applied to
+   * the k-vectors of the original cell (stored in the same order)
+   * The rotation matrices are generated randomly to avoid systematic
+   * biases.
+   */
+  BOOST_AUTO_TEST_CASE(kvecgen_rotation_test) {
+    // To make sure that rounding errors do not accidentally lead
+    // to unwanted errors, use a cutoff and cells that are quaranteed
+    // to be stable from Legendre's three squares theorem:
+    // For an integer (reciprocal) lattice or a sublattice thereof,
+    // the squared norm of k is an integer that is a sum of three
+    // squares. The numbers in the sequence https://oeis.org/A004215
+    // including the number 92 cannot be written as the sum of three
+    // squares, thus ensuring that there are no k-vectors whose squared
+    // norm satisfies 91 < k^2 93, ensuring stability with respect to
+    // the random rotation matrices as well as compiler choice etc.
+    // Note that the parameters are still chosen in a way that closely
+    // represent typical cells and cutoffs in applications.
+    double cutoff = sqrt(92);
+    std::vector<Eigen::Matrix3d> cells{};
+    Eigen::Matrix3d cell0;
+    // Cubic cell
+    cell0 << 1., 0., 0., 0., 1., 0., 0., 0., 1.;
+    cells.push_back(cell0);
+    // Tetragonal cell
+    cell0 << 2., 0., 0., 0., 1., 0., 0., 0., 1.;
+    cells.push_back(cell0);
+    // Orthorhombic cell
+    cell0 << 2., 0., 0., 0., 1., 0., 0., 0., 3.;
+    cells.push_back(cell0);
+
+    // Generate rotation matrices with fixed seed
+    srand(4653056);
+    size_t num_random = 100;
+    std::vector<Eigen::Matrix3d> rotation_matrices{};
+    for (size_t index = 0; index < num_random; index++) {
+      auto random_matrix = Eigen::Matrix3d::Random();
+      const auto rotation_matrix_0 =
```

Comment:
```suggestion
      Eigen::Matrix3d rotation_matrix_0 =
```
I think auto makes the type `const Eigen::HouseholderSequence<Eigen::Matrix<double, 3, 3, 0, 3, 3>, Eigen::Matrix<double, 3, 1, 0, 3, 1>, 1>`
I am not sure what then in the multiplication later happens, but it is probably not a matrix multiplication

Because auto is interpreted differently interpreted from interpret compiler, it can work on your machine probably using a gnu compiler.

Reply:
> Because auto is interpreted differently from different compiler, it can work on your machine using a gnu compiler while failing with lang

That's false. `auto` deduction rules are defined by the standard (they are the same rules applied during template substitution) and the same over different compilers (or this is a compiler bug). What might be happening here is that Eigen is doing strange things with templates, which auto take as they are; and using an explicit type forces a cast & evaluate the Eigen template. Overall, that's an Eigen issue more than a compiler/auto issue.